src/Agent.php

```diff
@@ -681,13 +681,99 @@ public static function getIcon()
+    {
+        global $PLUGIN_HOOKS;
+
+        // Actions to apply to the agent or item
+        /** @phpstan-var array{item_action: bool, callback: callable(?Agent, array, ?CommonDBTM):bool}[] $actions_to_apply */
+        $actions_to_apply = [];
+        $need_item = false;
+        if (isset($config['stale_agents_status']) && $config['stale_agents_status'] >= 0) {
+            // 0 = none, -1 or other negative = no change
+            $need_item = true;
+            $actions_to_apply[] = [
+                'item_action' => true,
+                'callback' => static function (?Agent $agent, array $config, ?CommonDBTM $item): bool {
+                    return $item !== null && $item->update([
+                        'id' => $item->fields['id'],
+                        'states_id' => $config['stale_agents_status']
+                    ]);
+                }
+            ];
+        }
+        $plugin_actions = $PLUGIN_HOOKS[Hooks::STALE_AGENT_CONFIG] ?? [];
+        /**
+         * @var string $plugin
+         * @phpstan-var array{label: string, item_action: boolean, render_callback: callable, action_callback: callable}[] $actions
+         */
+        foreach ($plugin_actions as $plugin => $actions) {
+            if (is_array($actions) && \Plugin::isPluginActive($plugin)) {
+                foreach ($actions as $action) {
+                    if ($action['item_action']) {
+                        $need_item = true;
+                    }
+                    $actions_to_apply[] = $action['action_callback'];
+                }
+            }
+        }
+        if (isset($config['stale_agents_clean']) && $config['stale_agents_clean']) {
+            $actions_to_apply[] = [
+                'item_action' => false,
+                'callback' => static function (?Agent $agent, array $config, ?CommonDBTM $item): bool {
+                    return $agent !== null && $agent->delete(['id' => $agent->fields['id']]);
+                }
+            ];
+        }
+        $item = null;
+
+        // If an action to apply needs the item
+        if ($need_item) {
+            $orphan = false;
+            try {
+                $item = new $agent->fields['itemtype']();
+                if ($item instanceof CommonDBTM) {
+                    if (!$item->getFromDB($agent->fields['items_id'])) {
+                        $orphan = true;
+                    }
+                } else {
+                    $orphan = true;
+                }
+            } catch (\Exception $e) {
+                $orphan = true;
+            }
+            if ($orphan) {
+                $item = null;
+            }
+        }
+
+        // Run all actions, with the clean action running last
+        foreach ($actions_to_apply as $action) {
+            // If the action needs the item, and we don't have one, skip it
+            if ($action['item_action'] && $item === null) {
+                continue;
+            }
```

Comment:
For a plugin developper that want to handle an action that acts on the agent itself, and optionnaly on related item, if one exists, current behaviour will require to declare 2 callables in plugin (the first one that will always be called, even when item is null, and the second one that will be called when item is not null). Not sure we will be able to document it clearly.
Also, plugins will probably want to act on item in most cases (I do not see any relevant use case of a plugin that would act only on the agent).
And, finally, a plugin that requires the item but that is declaring `'item_action' => false` may still receive the item in its callback, depending on GLPI conf and on other plugins, and so may luckily work in certain conditions on not on others.

IMHO, we could consider that `$need_item` is true as long as a plugin hook is declared. Plugin would have to check `$item !== null` in their callback, as you did in yours.
```suggestion
```

Reply:
`item_action` was intended to be a boolean indicating if the plugin's action works on the agent's related item instead of or in addition to the agent itself.

Reply:
Indeed, but you may want to work on both agent and its related item, and in this case, you probably want to work on agent even if related item is not available anymore.